src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,695 @@
+    
+    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE_REPLY, 
+                                  &reply_hdr, sizeof(reply_hdr), 
+                                  ucp_proto_am_zcopy_req_complete);
+}
+
+static ucs_status_t ucp_am_zcopy_multi(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_am_long_hdr_t hdr;
+    
+    hdr.ep         = ucp_request_get_dest_ep_ptr(req);
+    hdr.msg_id     = req->send.tag.message_id;
+    hdr.offset     = req->send.state.dt.offset;
+    hdr.am_id      = req->send.am.am_id;
+    hdr.total_size = req->send.length;
+    
+    return ucp_do_am_zcopy_multi(self, UCP_AM_ID_MULTI, 
+                                 UCP_AM_ID_MULTI,
+                                 &hdr, sizeof(hdr),
+                                 &hdr, sizeof(hdr),
+                                 ucp_proto_am_zcopy_req_complete, 0);
+}
+
+static ucs_status_t ucp_am_zcopy_multi_reply(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_am_long_hdr_t hdr;
+
+    hdr.ep         = ucp_request_get_dest_ep_ptr(req);
+    hdr.msg_id     = req->send.tag.message_id;
+    hdr.offset     = req->send.state.dt.offset;
+    hdr.am_id      = req->send.am.am_id;
+    hdr.total_size = req->send.length;
+    
+    return ucp_do_am_zcopy_multi(self, UCP_AM_ID_MULTI_REPLY, 
+                                 UCP_AM_ID_MULTI_REPLY,
+                                 &hdr, sizeof(hdr),
+                                 &hdr, sizeof(hdr),
+                                 ucp_proto_am_zcopy_req_complete, 0);
+}
+
+static void ucp_am_send_req_init(ucp_request_t *req, ucp_ep_h ep,
+                                 const void *buffer, uintptr_t datatype,
+                                 size_t count, uint16_t flags, 
+                                 uint16_t am_id)
+{
+    req->flags          = UCP_REQUEST_FLAG_SEND_AM;
+    req->send.ep        = ep;
+    req->send.am.am_id  = am_id;
+    req->send.am.flags  = flags;
+    req->send.buffer    = (void *) buffer;
+    req->send.datatype  = datatype;
+    req->send.mem_type  = UCT_MD_MEM_TYPE_HOST;
+    req->send.lane      = ep->am_lane;
+
+    ucp_request_send_state_init(req, datatype, count);
+    req->send.length = ucp_dt_length(req->send.datatype, count,
+                                     req->send.buffer,
+                                     &req->send.state.dt);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_am_send_req(ucp_request_t *req, size_t count,
+                const ucp_ep_msg_config_t * msg_config,
+                ucp_send_callback_t cb, const ucp_proto_t *proto)
+{
+    
+    size_t zcopy_thresh = ucp_proto_get_zcopy_threshold(req, msg_config,
+                                                        count, SIZE_MAX);
+    size_t max_short;
+    ucs_status_t status;
+    
+    max_short = ucp_proto_get_short_max(req, msg_config);
```

Comment:
don't change ucp_proto_get_short_max(), we can decide here to avoid short protocol